query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/PostgresClientFactory.java

```diff
@@ -131,5 +178,42 @@ static String resolveStatement(String query, Params params) {
     private static String getStringParam(String value) {
       return "'" + value.replace("'", "''") + "'";
     }
+
+    @SneakyThrows
+    Observable<Row> convert(ResultSet resultSet) {
+      List<Row.Builder> rowBuilderList = new ArrayList<>();
+      while (resultSet.next()) {
+        Row.Builder builder = Row.newBuilder();
+        rowBuilderList.add(builder);
+        ResultSetMetaData metaData = resultSet.getMetaData();
+        int columnCount = metaData.getColumnCount();
+        if (columnCount > 0) {
+          for (int c = 1; c <= columnCount; c++) {
+            int colType = metaData.getColumnType(c);
+            Value convertedColVal;
+            if (colType == Types.ARRAY) {
+              Array colVal = resultSet.getArray(c);
+              convertedColVal =
+                  Value.newBuilder()
+                      .setValueType(ValueType.STRING)
+                      .setString(
+                          MAPPER.writeValueAsString(
+                              colVal != null ? colVal.getArray() : Collections.emptyList()))
+                      .build();
+            } else {
+              String colVal = resultSet.getString(c);
+              convertedColVal =
+                  colVal != null
+                      ? Value.newBuilder().setValueType(ValueType.STRING).setString(colVal).build()
+                      : NULL_VALUE;
+            }
+            builder.addColumn(convertedColVal);
+          }
+        }
+      }
+      return Observable.fromIterable(rowBuilderList)
```

Comment:
I love observables as much (more) as the next guy, but this observable isn't being used for anything async, so what's the purpose? If there's a reason to use Rx, should start with the actual db call and make the whole thing observable rather than just doing everything sync and wrapping the result.

Reply:
Caller requires an observable to be returned. However you are right current implementation is not async.